File: ios-native/Sources/SportkalenderiOS/Domain/TvRights.swift
import Foundation

private struct Rule {
    let weekday: Int
    let hour: Int
    let minute: Int
    let channels: [TvChannel]
}

private let rules: [String: [Rule]] = [
    "bl1": [
        Rule(weekday: 6, hour: 20, minute: 30, channels: [TvChannel(name: "Sky", isFree: false)]),
        Rule(weekday: 7, hour: 15, minute: 30, channels: [TvChannel(name: "Sky", isFree: false), TvChannel(name: "DAZN Konf.", isFree: false)]),
        Rule(weekday: 7, hour: 18, minute: 30, channels: [TvChannel(name: "Sky", isFree: false)]),
        Rule(weekday: 1, hour: 15, minute: 30, channels: [TvChannel(name: "DAZN", isFree: false)]),
        Rule(weekday: 1, hour: 17, minute: 30, channels: [TvChannel(name: "DAZN", isFree: false)]),
        Rule(weekday: 1, hour: 19, minute: 30, channels: [TvChannel(name: "DAZN", isFree: false)])
    ],
    "bl2": [
        Rule(weekday: 6, hour: 18, minute: 30, channels: [TvChannel(name: "Sky", isFree: false)]),
        Rule(weekday: 7, hour: 13, minute: 0, channels: [TvChannel(name: "Sky", isFree: false)]),
        Rule(weekday: 7, hour: 20, minute: 30, channels: [TvChannel(name: "Sky", isFree: false), TvChannel(name: "RTL", isFree: true)]),
        Rule(weekday: 1, hour: 13, minute: 30, channels: [TvChannel(name: "Sky", isFree: false)])
    ],
    "dfb": [
        Rule(weekday: -1, hour: -1, minute: -1, channels: [TvChannel(name: "Sky", isFree: false), TvChannel(name: "ARD/ZDF", isFree: true)])
    ],
    "ucl": [
        Rule(weekday: -1, hour: -1, minute: -1, channels: [TvChannel(name: "DAZN", isFree: false), TvChannel(name: "Prime", isFree: false), TvChannel(name: "ZDF", isFree: true)])
    ]
]

func resolveTvChannels(league: String, kickoff: Date?) -> [TvChannel] {
    guard let kickoff, let leagueRules = rules[league] else { return [] }
    let cal = Calendar(identifier: .gregorian)
    let comps = cal.dateComponents(in: TimeZone(identifier: "Europe/Berlin")!, from: kickoff)
    let weekday = comps.weekday ?? -1
    let hour = comps.hour ?? -1
    let minute = comps.minute ?? -1

    if let exact = leagueRules.first(where: { $0.weekday == weekday && $0.hour == hour && $0.minute == minute }) {
        return exact.channels
    }
    return leagueRules.first(where: { $0.weekday == -1 })?.channels ?? []
}

File: android-native/app/src/main/java/de/sportkalender/nativeapp/domain/TvRights.kt
package de.sportkalender.nativeapp.domain

import de.sportkalender.nativeapp.model.TvChannel
import java.time.ZonedDateTime

private data class Rule(
    val day: Int,
    val hour: Int,
    val minute: Int,
    val channels: List<TvChannel>
)

private val rightsByLeague: Map<String, List<Rule>> = mapOf(
    "bl1" to listOf(
        Rule(5, 20, 30, listOf(TvChannel("Sky", false))),
        Rule(6, 15, 30, listOf(TvChannel("Sky", false), TvChannel("DAZN Konf.", false))),
        Rule(6, 18, 30, listOf(TvChannel("Sky", false))),
        Rule(0, 15, 30, listOf(TvChannel("DAZN", false))),
        Rule(0, 17, 30, listOf(TvChannel("DAZN", false))),
        Rule(0, 19, 30, listOf(TvChannel("DAZN", false))),
        Rule(2, 20, 30, listOf(TvChannel("Sky", false))),
        Rule(3, 20, 30, listOf(TvChannel("Sky", false)))
    ),
    "bl2" to listOf(
        Rule(5, 18, 30, listOf(TvChannel("Sky", false))),
        Rule(6, 13, 0, listOf(TvChannel("Sky", false))),
        Rule(6, 20, 30, listOf(TvChannel("Sky", false), TvChannel("RTL", true))),
        Rule(0, 13, 30, listOf(TvChannel("Sky", false)))
    ),
    "dfb" to listOf(
        Rule(-1, -1, -1, listOf(TvChannel("Sky", false), TvChannel("ARD/ZDF", true)))
    ),
    "ucl" to listOf(
        Rule(-1, -1, -1, listOf(TvChannel("DAZN", false), TvChannel("Prime", false), TvChannel("ZDF", true)))
    )
)

fun resolveTvChannels(league: String, kickoff: ZonedDateTime?): List<TvChannel> {
    if (kickoff == null) return emptyList()
    val rules = rightsByLeague[league] ?: return emptyList()
    val weekday = kickoff.dayOfWeek.value % 7
    val hour = kickoff.hour
    val minute = kickoff.minute

    val exact = rules.firstOrNull { it.day == weekday && it.hour == hour && it.minute == minute }
    if (exact != null) return exact.channels

    val wildcard = rules.firstOrNull { it.day == -1 }
    return wildcard?.channels.orEmpty()
}
